src/executor/copy_from_file.rs

```diff
@@ -104,27 +100,33 @@ impl CopyFromFileExecutor {
                         actual: record.len(),
                     });
                 }
-                for ((s, builder), ty) in record
-                    .iter()
-                    .zip(&mut array_builders)
-                    .zip(&self.plan.logical().column_types().to_vec())
-                {
-                    if !ty.is_nullable() && s.is_empty() {
-                        return Err(ExecutorError::NotNullable);
-                    }
-                    builder.push_str(s)?;
+
+                let str_row_data: Result<Vec<&str>, _> =
+                    izip!(record.iter(), self.plan.logical().column_types())
+                        .map(|(v, ty)| {
+                            if !ty.is_nullable() && v.is_empty() {
+                                return Err(ExecutorError::NotNullable);
+                            }
+                            Ok(v)
+                        })
+                        .collect();
+
+                // push a str row and parse
+                if let Some(e) = chunk_builder.push_str_row(str_row_data?).err() {
+                    return Err(ExecutorError::Convert(e));
                 }
             }
+
             // update progress bar
             bar.set_position(iter.reader().position().byte());
 
             // send data chunk
-            let chunk: DataChunk = array_builders.into_iter().collect();
-
             #[allow(clippy::collapsible_if)]
-            if chunk.cardinality() > 0 {
-                if token.is_cancelled() || tx.blocking_send(chunk).is_err() {
-                    return Err(ExecutorError::Abort);
+            if let Some(chunk) = { chunk_builder }.take() {
+                if chunk.cardinality() > 0 {
```

Comment:
From the doc of `DataChunkBuilder::take`: If there are no remaining rows, `None` will be returned.

So I guess `if chunk.cardinality() > 0 {` can be removed. (And maybe also `#[allow(clippy::collapsible_if)]`?)